### DISCUSSION TITLE
Manual BPM input for Rate Adjust mods

### DISCUSSION THREAD

**Discussion by rozukke** [ppy/osu#22616] [Ideas]
During discussions in today's community meeting, multiple people mentioned the difference between rate adjust and BPM adjust in regards to changing the speed of beatmaps. A simple way of implementing this distinction in a relatively user friendly way would be a text input for a target BPM. This is basically 100% a UI change but there's still some considerations, like what 'BPM' means for variable BPM maps. A basic form of it would look something like this:

![image](https://user-images.githubusercontent.com/73006620/218296810-f0a577a0-80ec-4cb9-a5c1-97744509cc6a.png)

In this example, the map BPM is 200 and typing 230 into the textbox brings the multiplier up to 1.15x, rounded to the closest .01x.

While this is basically an entirely superficial change if implemented as above, this would require the mod overlay to have access to the current beatmap (which afaik is not the case right now).

Additionally, for variable BPM maps it would have to use the most common BPM (i.e. 'mostly X bpm' in the beatmap details).

Any thoughts/concerns/suggestions appreciated.

**Comment by Axilotl17** [comment]
There is a discrepancy between how the skinning BPM is managed and how you theorize this would work; 
For the BPM display while in the carousel menu, the BPM value is the BPM of the first timing point. 
I personally think they should both use the most common BPM, as you suggested. However for now it might be best to follow current precedent?

**Reply by rozukke** [reply]
The first timing point in a lot of cases is not representative of the actual speed of the map with many songs having slow intro sections/BPM speedups a relatively short time into the map. It also doesn't sound like the current BPM display implementation is ideal since I'd assume it should display the BPM of the currently playing bit of music.

**Reply by bdach** [reply]
> For the BPM display while in the carousel menu, the BPM value is the BPM of the first timing point

huh? this is false. where are you getting this from?

see

https://github.com/ppy/osu/blob/8fb72b971ddb236bd9ea2f6fef6a93a16d82da61/osu.Game/Screens/Select/BeatmapInfoWedge.cs#L420-L433

> It also doesn't sound like the current BPM display implementation is ideal since I'd assume it should display the BPM of the currently playing bit of music.

the skin bpm display does this already.

**Reply by Axilotl17** [reply]
I swear when I tested it, it was first timing point but i just tried again and yeah you're totally right. I have no idea how i got that.

maybe because even though most of the **audio** was *x* bpm, most of the **drain time** was *y* bpm?

**Reply by Axilotl17** [reply]
>maybe because even though most of the audio was x bpm, most of the drain time was y bpm?

nope, even with this they line up. I guess I made a mistake. sorry!

**Comment by bdach** [comment]
as I said on discord already I believe this would be better addressed by displaying the bpm post-mod application (which is already planned) rather than adding an extra textbox for this.

**Reply by Axilotl17** [reply]
but i find myself calculating the numbers more often than not. having perfect bpm with DT is so helpful sometimes, messing with the slider to find the perfect bpm value just isnt usable from a player perspective.

**Reply by monochrome22** [reply]
With just displaying the BPM, it would mean you need to adjust the slider every time you change the map to get the BPM you want. 

Manual BPM input would look ugly, but I don't think it would be confusing. If a player sets the BPM at 250BPM, then selects a map that is 100BPM, it would only go to 200BPM. If the player becomes confused and they start wondering why is it 200BPM instead of 250BPM, their first thought would be to check the mod, and they would see the mod is at 2.00x. At that point they should immediately realize it's only 200BPM because the speed increase can't go any higher. It might help to also display somewhere "The BPM of the selected map can't be set to 250", maybe below the BPM box in the mod setting. 

It might be best to add a toggle "Only display beatmaps that can be set to this BPM". So if a player sets the BPM at 240, it would only display maps that have base BPM 120-238. This shouldn't be enabled without a toggle, because then a player might search for a map and they would become confused if it doesn't show. 
The presence of this toggle would also help with having the players understand that not all maps can be set to a certain BPM.

**Comment by zabbazebra** [comment]
100% agree. Would greatly increase quality of life of the player

**Reply by Givikap120** [reply]
It won't be used much because it will most likely result in uneven rate, what will make the score unranked
